File: kakao.app/js/ked/view/status/status.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, isStr: function (o) {return typeof o === 'string' || o instanceof String}, rpad: function (l,s='',c=' ') {s=String(s); while(s.length<l){s+=c} return s}, clamp: function (l,h,v) { var ll = Math.min(l,h), hh = Math.max(l,h); if (!_k_.isNum(v)) { v = ll }; if (v < ll) { v = ll }; if (v > hh) { v = hh }; if (!_k_.isNum(v)) { v = ll }; return v }, k: { f:(r,g,b)=>'\x1b[38;5;'+(16+36*r+6*g+b)+'m', F:(r,g,b)=>'\x1b[48;5;'+(16+36*r+6*g+b)+'m', r:(i)=>(i<6)&&_k_.k.f(i,0,0)||_k_.k.f(5,i-5,i-5), R:(i)=>(i<6)&&_k_.k.F(i,0,0)||_k_.k.F(5,i-5,i-5), g:(i)=>(i<6)&&_k_.k.f(0,i,0)||_k_.k.f(i-5,5,i-5), G:(i)=>(i<6)&&_k_.k.F(0,i,0)||_k_.k.F(i-5,5,i-5), b:(i)=>(i<6)&&_k_.k.f(0,0,i)||_k_.k.f(i-5,i-5,5), B:(i)=>(i<6)&&_k_.k.F(0,0,i)||_k_.k.F(i-5,i-5,5), y:(i)=>(i<6)&&_k_.k.f(i,i,0)||_k_.k.f(5,5,i-5), Y:(i)=>(i<6)&&_k_.k.F(i,i,0)||_k_.k.F(5,5,i-5), m:(i)=>(i<6)&&_k_.k.f(i,0,i)||_k_.k.f(5,i-5,5), M:(i)=>(i<6)&&_k_.k.F(i,0,i)||_k_.k.F(5,i-5,5), c:(i)=>(i<6)&&_k_.k.f(0,i,i)||_k_.k.f(i-5,5,5), C:(i)=>(i<6)&&_k_.k.F(0,i,i)||_k_.k.F(i-5,5,5), w:(i)=>'\x1b[38;5;'+(232+(i-1)*3)+'m', W:(i)=>'\x1b[48;5;'+(232+(i-1)*3+2)+'m', wrap:(open,close,reg)=>(s)=>open+(~(s+='').indexOf(close,4)&&s.replace(reg,open)||s)+close, F256:(open)=>_k_.k.wrap(open,'\x1b[39m',new RegExp('\\x1b\\[39m','g')), B256:(open)=>_k_.k.wrap(open,'\x1b[49m',new RegExp('\\x1b\\[49m','g'))}, isNum: function (o) {return !isNaN(o) && !isNaN(parseFloat(o)) && (isFinite(o) || o === Infinity || o === -Infinity)}};_k_.y5=_k_.k.F256(_k_.k.y(5));_k_.w3=_k_.k.F256(_k_.k.w(3))

var fileposSyntax, status

import kxk from "../../../kxk.js"
let post = kxk.post
let slash = kxk.slash
let kstr = kxk.kstr
let kseg = kxk.kseg

import nfs from "../../../kxk/nfs.js"

import belt from "../../edit/tool/belt.js"

import filepos from "../../edit/mode/filepos.js"

import color from "../../theme/color.js"
import theme from "../../theme/theme.js"

import view from "../base/view.js"
import crumbs from "../base/crumbs.js"
import bubble from "../base/bubble.js"

import statusfile from "./statusfile.js"


fileposSyntax = (function ()
{
    function fileposSyntax ()
    {
        this["getColor"] = this["getColor"].bind(this)
        this.color = {}
        this.color.number = theme.status.filepos
        this.color.symbol = color.darken(this.color.number,0.8)
    }

    fileposSyntax.prototype["getColor"] = function (x, y, ch)
    {
        switch (ch)
        {
            case '':
                return this.color.symbol

            default:
                return this.color.number
        }

    }

    return fileposSyntax
})()


status = (function ()
{
    _k_.extend(status, view)
    function status (screen, editor)
    {
        this.screen = screen
        this.editor = editor
    
        this["draw"] = this["draw"].bind(this)
        this["layout"] = this["layout"].bind(this)
        this["setFile"] = this["setFile"].bind(this)
        this["onMouse"] = this["onMouse"].bind(this)
        this["onFileAction"] = this["onFileAction"].bind(this)
        this["droopCrumb"] = this["droopCrumb"].bind(this)
        this["onCrumbsAction"] = this["onCrumbsAction"].bind(this)
        this["onFileposAction"] = this["onFileposAction"].bind(this)
        this["onStatusFilepos"] = this["onStatusFilepos"].bind(this)
        status.__super__.constructor.call(this,this.screen,'status')
        this.state = this.editor.state
        this.gutter = 4
        this.file = ''
        this.drawTime = ''
        this.fileposl = []
        this.fileoffs = []
        this.pointerType = 'pointer'
        this.setColor('gutter',theme.gutter.bg)
        this.crumbs = new crumbs(this.screen,'status_crumbs')
        this.statusfile = new statusfile(this.screen,'status_file')
        this.filepos = new bubble(this.screen,"status_filepos")
        this.crumbs.setColor('empty_left',this.color.gutter)
        this.crumbs.setColor('empty_right',theme.status.empty)
        this.filepos.setColor('empty',theme.status.empty)
        this.filepos.syntax = new fileposSyntax
        this.crumbs.on('action',this.onCrumbsAction)
        this.statusfile.on('action',this.onFileAction)
        this.filepos.on('action',this.onFileposAction)
        post.on('status.filepos',this.onStatusFilepos)
    }

    status.prototype["onStatusFilepos"] = function (fileposl, fileoffs)
    {
        this.fileposl = fileposl
        this.fileoffs = fileoffs
    
        if (this.fileposl.length > 1)
        {
            return this.filepos.set({tilde:`${(this.fileoffs ? this.fileposl.length - this.fileoffs - 1 : '')}${this.fileposl.length - 1}`})
        }
        else
        {
            return this.filepos.set(null)
        }
    }

    status.prototype["onFileposAction"] = function (action, item)
    {
        var files

        switch (action)
        {
            case 'leave':
                console.log('hide filepos files?')
                break
            case 'click':
                return post.emit('filepos.swapPrevious')

            case 'enter':
                if (filepos.fileposl.length > 1)
                {
                    files = filepos.fileposl.map(function (fp)
                    {
                        return fp[0]
                    })
                    files = files.reverse()
                    if (filepos.offset === 0)
                    {
                        files.shift()
                    }
                    return post.emit('droop.show',{files:files,pos:[this.filepos.cells.x + parseInt(this.filepos.cells.cols / 2),this.filepos.cells.y + 1]})
                }
                break
        }

    }

    status.prototype["onCrumbsAction"] = function (action, path, event)
    {
        switch (action)
        {
            case 'click':
                if (!_k_.empty(event.mods))
                {
                    return post.emit('dircol.root',path)
                }
                else
                {
                    return post.emit('browse.dir',path)
                }
                break
            case 'enter':
                return this.droopCrumb(path,event)

        }

    }

    status.prototype["droopCrumb"] = async function (path, crumb)
    {
        var files, x

        path = slash.untilde(path)
        files = await nfs.list(path,{recursive:false})
        x = this.crumbs.cells.x + parseInt((crumb.cols[1] + crumb.cols[0]) / 2)
        return post.emit('droop.show',{files:files,pos:[x,this.crumbs.cells.y + 1]})
    }

    status.prototype["onFileAction"] = function (action, file)
    {
        switch (action)
        {
            case 'click':
                return post.emit('browse.dir',slash.dir(file),file)

        }

    }

    status.prototype["onMouse"] = function (event)
    {
        var col, cret, fret, row, sret

        cret = this.crumbs.onMouse(event)
        sret = this.statusfile.onMouse(event)
        fret = this.filepos.onMouse(event)
        if (sret || cret || fret)
        {
            return sret || cret || fret
        }
        status.__super__.onMouse.call(this,event)
        var _a_ = this.eventPos(event); col = _a_[0]; row = _a_[1]

        if (this.hover)
        {
            post.emit('pointer',this.pointerType)
        }
        switch (event.type)
        {
            case 'press':
                if (this.hover)
                {
                    if ((0 <= col && col < 4))
                    {
                        post.emit('dircol.toggle')
                        return {redraw:true}
                    }
                    if ((this.cells.cols - 12 <= col && col < this.cells.cols))
                    {
                        post.emit('funcol.toggle')
                        return {redraw:true}
                    }
                }
                break
        }

        return this.hover
    }

    status.prototype["setFile"] = function (file)
    {
        this.file = file
    
        if (_k_.empty(this.file))
        {
            return
        }
        this.crumbs.set(slash.dir(this.file))
        return this.statusfile.set(this.file)
    }

    status.prototype["layout"] = function (x, y, w, h)
    {
        status.__super__.layout.call(this,x,y,w,h)
    
        var cw

        cw = parseInt(w / 2)
        x += this.gutter + 1
        this.crumbs.layout(x,y,cw,1)
        this.crumbs.layout(x,y,this.crumbs.rounded.length,1)
        x += this.crumbs.rounded.length
        this.statusfile.layout(x,y,this.statusfile.rounded.length,1)
        x += this.statusfile.rounded.length
        return this.filepos.layout(x,y,this.filepos.rounded.length,1)
    }

    status.prototype["draw"] = function ()
    {
        var add, ch, ci, colno, cols, cur, cursor, dty, fg, fnl, hil, i, rdo, sel, set, x, y

        if (this.hidden() || _k_.empty(this.file))
        {
            return
        }
        x = 0
        y = 0
        cursor = this.state.mainCursor()
        cols = this.cells.cols
        fnl = this.file.length
        rdo = this.state.hasRedo()
        dty = this.state.isDirty()
        set = (function (x, char, fg, bg)
        {
            if (_k_.isStr(fg))
            {
                fg = theme.status[fg]
            }
            if (_k_.isStr(bg))
            {
                bg = theme.status[bg]
            }
            this.cells.set(x,y,char,fg,bg)
            return 1
        }).bind(this)
        add = (function (char, fg, bg)
        {
            return x += set(x,char,fg,bg)
        }).bind(this)
        add('','col',this.color.gutter)
        colno = _k_.rpad(this.gutter - 1,`${cursor[0]}`)
        for (var _a_ = ci = 1, _b_ = this.gutter; (_a_ <= _b_ ? ci < this.gutter : ci > this.gutter); (_a_ <= _b_ ? ++ci : --ci))
        {
            fg = (cursor[0] ? 'fg' : 'col_zero')
            if (belt.isLinesPosOutside(this.state.s.lines,cursor))
            {
                fg = 'col_empty'
            }
            add(((ci - 1 < colno.length) ? colno[ci - 1] : ' '),fg,'col')
        }
        add('','col',this.color.gutter)
        this.crumbs.draw()
        this.statusfile.draw()
        this.filepos.draw()
        x += this.crumbs.rounded.length
        x += this.statusfile.rounded.length
        x += this.filepos.rounded.length
        add('','dark','empty')
        if (dty)
        {
            add('','dirty','dark')
        }
        if (dty)
        {
            add(' ','dirty','dark')
        }
        if (rdo)
        {
            add('','redo','dark')
        }
        add(' ','dark','dark')
        if (this.state.s.cursors.length > 1)
        {
            cur = `${this.state.s.cursors.length}♦`
            for (var _c_ = i = 0, _d_ = cur.length; (_c_ <= _d_ ? i < cur.length : i > cur.length); (_c_ <= _d_ ? ++i : --i))
            {
                add(cur[i],((i < cur.length - 1) ? 'cur' : color.darken(theme.status.cur)),'dark')
            }
        }
        if (this.state.s.selections.length)
        {
            sel = `${this.state.s.selections.length}≡`
            for (var _e_ = i = 0, _f_ = sel.length; (_e_ <= _f_ ? i < sel.length : i > sel.length); (_e_ <= _f_ ? ++i : --i))
            {
                add(sel[i],((i < sel.length - 1) ? 'sel' : color.darken(theme.status.sel)),'dark')
            }
        }
        if (this.state.s.highlights.length)
        {
            hil = `${this.state.s.highlights.length}❇`
            for (var _10_ = i = 0, _11_ = hil.length; (_10_ <= _11_ ? i < hil.length : i > hil.length); (_10_ <= _11_ ? ++i : --i))
            {
                add(hil[i],((i < hil.length - 1) ? 'hil' : color.darken(theme.status.hil)),'dark')
            }
        }
        for (var _12_ = ci = x, _13_ = cols - 1; (_12_ <= _13_ ? ci < cols - 1 : ci > cols - 1); (_12_ <= _13_ ? ++ci : --ci))
        {
            add(' ',null,'dark')
        }
        add('','dark',this.color.gutter)
        ci = _k_.clamp(0,3,parseInt((this.time / (1000 * 1000) - 8) / 8))
        ch = ' •'[ci]
        fg = [[32,32,32],[0,96,0],[255,0,0],[255,255,0]][ci]
        switch (ch)
        {
            case '•':
                1
                break
            case '':
                1
                break
            case '':
                console.log(`${_k_.y5(ch)} ${_k_.w3(kstr.time(BigInt(this.time)))}`)
                break
        }

        return set(cols - 2,ch,fg,'dark')
    }

    return status
})()

export default status;

File: kakao.app/love/lua/view/editor/status.lua
--[[
     ███████  █████████   ███████   █████████  ███   ███   ███████  
    ███          ███     ███   ███     ███     ███   ███  ███       
    ███████      ███     █████████     ███     ███   ███  ███████   
         ███     ███     ███   ███     ███     ███   ███       ███  
    ███████      ███     ███   ███     ███      ███████   ███████   

    the row above the fileeditor
    displays currently edited file and editor state information 
--]]

view = require "view.base.view"
crumbs = require "view.base.crumbs"
bubble = require "view.base.bubble"
statusfile = require "view.editor.statusfile"


local fileposSyntax = class("fileposSyntax")
    


function fileposSyntax:init() 
        self.color = {}
        self.color.number = theme.status.filepos
        -- @color.symbol = color.darken @color.number 0.8
        self.color.symbol = theme.status.filepos
        return self
    end


function fileposSyntax:getColor(x, y, ch) 
        if (ch == '') then 
    return self.color.symbol
        else 
    return self.color.number
        end
    end


local status = class("status", view)
    


function status:init(editor) 
        self.editor = editor
        
        view.init(self, 'status')
        
        self.state = self.editor.state
        self.gutter = 4
        self.file = ''
        self.drawTime = ''
        self.fileposl = array()
        self.fileoffs = array()
        
        self.pointerType = 'pointer'
        
        self:setColor('gutter', theme.gutter.bg)
        
        self.crumbs = crumbs("status_crumbs")
        self.statusfile = statusfile("status_file")
        self.filepos = bubble("status_filepos")
        
        self.crumbs:setColor('empty_left', self.color.gutter)
        self.crumbs:setColor('empty_right', theme.status.empty)
        
        self.filepos:setColor('empty', theme.status.empty)
        self.filepos.syntax = fileposSyntax()
        
        self.crumbs:on('action', self.onCrumbsAction, self)
        self.statusfile:on('action', self.onFileAction, self)
        self.filepos:on('action', self.onFileposAction, self)
        
        -- post∙on 'status.filepos' @onStatusFilepos @
        return self
    end


function status:onStatusFilepos(fileposl, fileoffs) 
        print("STATUS FILEPOS", fileposl)
        print("STATUS FILEOFF", fileoffs)
        
        self.fileposl = fileposl
        self.fileoffs = fileoffs
        
        if (self.fileposl:len() > 1) then 
            local off = (function () 
    if self.fileoffs then 
    return ((self.fileposl:len() - self.fileoffs) - 1) else 
    return ''
                  end
end)()
            local tilde = "" .. tostring(tilde) .. "" .. tostring((self.fileposl:len() - 1)) .. ""
            self.filepos:set({tilde = tilde})
        else 
            self.filepos:set(nil)
        end
        
        return self
    end


function status:onFileposAction(action, item) 
        if (action == 'leave') then 
    return print('hide filepos files?')
        elseif (action == 'click') then 
    return post:emit('filepos.swapPrevious')
        elseif (action == 'enter') then 
                if (#filepos.fileposl > 1) then 
                    local files = filepos.fileposl:map(function (fp) 
    return fp[0]
end)
                    -- files = files.reverse()
                    if (filepos.offset == 0) then files:shift() end
                    return -- post∙emit 'droop.show' files:files pos:[@filepos.cells.x+int(@filepos.cells.cols/2) @filepos.cells.y+1]
                end
        end
    end


function status:onCrumbsAction(action, path, event) 
        -- switch action
        -- 
        --     'click'  
        --         
        --         clearTimeout @droopTimer
        --         if valid event.mods
        --             post∙emit 'dircol.root' path
        --         else
        --             post∙emit 'browse.dir' path
        --             
        --     'enter'
        --         
        --         post∙emit 'droop.hide'
        --         clearTimeout @droopTimer
        --         @droopTimer = setTimeout ((p e) -> ->  @droopCrumb(p e))(path event), 500
        --         
        --     'leave'
        --         
        --         clearTimeout @droopTimer
        return --         delete @droopTimer
    end


function status:droopCrumb(path, crumb) 
        -- clearTimeout @droopTimer
        -- delete @droopTimer
        
        path = slash.untilde(path)
        local files = slash.walk(path, {recursive = false})
        local x = (self.crumbs.cells.x + int(((crumb.cols[2] + crumb.cols[1]) / 2)))
        return post:emit('droop.show', {files = files, pos = array(x, (self.crumbs.cells.y + 1))})
    end


function status:onFileAction(action, file) 
        if empty(file) then return end
        print("onFileAction", action, file)
        if (action == 'click') then 
    return post:emit('browse.dir', slash.dir(file), file)
        end
    end

-- 00     00   0000000   000   000   0000000  00000000  
-- 000   000  000   000  000   000  000       000       
-- 000000000  000   000  000   000  0000000   0000000   
-- 000 0 000  000   000  000   000       000  000       
-- 000   000   0000000    0000000   0000000   00000000  


function status:onMouse(event) 
        local cret = self.crumbs:onMouse(event)
        local sret = self.statusfile:onMouse(event)
        local fret = self.filepos:onMouse(event)
        
        if ((sret or cret) or fret) then 
            return ((sret or cret) or fret)
        end
        
        view.onMouse(self, event)
        
        local col, row = unpack(self:eventPos(event))
        
        if self.hover then 
            post:emit('pointer', self.pointerType)
        end
        
        if (event.type == 'press') then 
                if self.hover then 
                    if ((1 <= col) and (col <= 4)) then 
                        post:emit('dircol.toggle')
                        return true
                    end
                    
                    if (((self.cells.cols - 12) <= col) and (col <= self.cells.cols)) then 
                        post:emit('funcol.toggle')
                        return true
                    end
                end
        end
        
        return self.hover
    end

--  0000000  00000000  000000000        00000000  000  000      00000000  
-- 000       000          000           000       000  000      000       
-- 0000000   0000000      000           000000    000  000      0000000   
--      000  000          000           000       000  000      000       
-- 0000000   00000000     000           000       000  0000000  00000000  


function status:setFile(file) 
        self.file = file
        if empty(self.file) then return end
        self.crumbs:set(slash.dir(self.file))
        return self.statusfile:set(self.file)
    end

-- ███       ███████   ███   ███   ███████   ███   ███  █████████
-- ███      ███   ███   ███ ███   ███   ███  ███   ███     ███   
-- ███      █████████    █████    ███   ███  ███   ███     ███   
-- ███      ███   ███     ███     ███   ███  ███   ███     ███   
-- ███████  ███   ███     ███      ███████    ███████      ███   


function status:layout(x, y, w, h) 
        view.layout(self, x, y, w, h)
        
        local cw = floor((w / 2))
        
        x = x + ((self.gutter + 1))
        
        self.crumbs:layout(x, y, cw, 1) -- 𝜏𝖍𝚒𝖘 ⟅◯⊚𝛋𝖘 ϝ𝚒𝖘𝖍𝛾 
        self.crumbs:layout(x, y, #self.crumbs.rounded, 1) -- ⟒𝖍𝛾 𝜏⟒⊚ ⟅𝕒𝛾◯𝓊𝜏𝖘?
        
        x = x + (#self.crumbs.rounded)
        
        self.statusfile:layout(x, y, #self.statusfile.rounded, 1)
        
        x = x + (#self.statusfile.rounded)
        
        return self.filepos:layout(x, y, #self.filepos.rounded, 1)
    end

-- 0000000    00000000    0000000   000   000  
-- 000   000  000   000  000   000  000 0 000  
-- 000   000  0000000    000000000  000000000  
-- 000   000  000   000  000   000  000   000  
-- 0000000    000   000  000   000  00     00  


function status:draw() 
        if self:hidden() then return end --or empty @file
        
        self.gutter = self.gutter or 4
        
        local x = 1
        local y = 1
        local cursor = self.state:mainCursor()
        local cols = self.cells.cols
        local fnl = #self.file
        local rdo = self.state:hasRedo()
        local dty = self.state:isDirty()
        local fg = array(255, 0, 0)
        local bg = array(55, 55, 55)
        
        
        function set(x, char, fg, bg) 
            if is(fg, "string") then fg = theme.status[fg] end
            if is(bg, "string") then bg = theme.status[bg] end
            self.cells:set(x, y, char, fg, bg)
            return 1
        end
        
        
        function add(char, fg, bg) 
    x = x + (set(x, char, fg, bg))
    return x
        end
        
        add('', 'col', self.color.gutter)
        local colno = kstr.rpad((self.gutter - 1), "" .. tostring(cursor[1]) .. "")
        for ci = 1, self.gutter-1 do 
            fg = (function () 
    if (cursor[1] > 1) then 
    return 'fg' else 
    return 'col_zero'
                 end
end)()
            if belt.isLinesPosOutside(self.state.s.lines, cursor) then 
                fg = 'col_empty'
            end
            
            local char = (function () 
    if ((ci - 1) < #colno) then 
    return string.sub(colno, (ci - 1), (ci - 1)) else 
    return ' '
                   end
end)()
            add(char, fg, 'col')
        end
        
        add('', 'col', self.color.gutter)
        
        self.crumbs:draw()
        self.statusfile:draw()
        self.filepos:draw()
        
        x = x + (#self.crumbs.rounded)
        x = x + (#self.statusfile.rounded)
        x = x + (#self.filepos.rounded)
        
        add('', 'dark', 'empty')
        if dty then add(' ', 'dirty', 'dark') end
        if dty then add(' ', 'dirty', 'dark') end
        if rdo then add('➜', 'redo', 'dark') end
        add(' ', 'dark', 'dark')
        
        if (self.state.s.cursors:len() > 1) then 
            local cur = "" .. tostring(self.state.s.cursors:len()) .. "♦"
            
            for i in iter(1, cur:len()) do 
                local color = (function () 
    if (i < cur:len()) then 
    return 'cur' else 
    return color.darken(theme.status.cur)
                        end
end)()
                add(cur[i], color, 'dark')
            end
        end
        
        if self.state.s.selections:len() then 
            local sel = "" .. tostring(self.state.s.selections:len()) .. "≡"
            
            for i in iter(1, sel:len()) do 
                local color = (function () 
    if (i < sel:len()) then 
    return 'sel' else 
    return color.darken(theme.status.sel)
                        end
end)()
                add(sel[i], color, 'dark')
            end
        end
        
        if self.state.s.highlights:len() then 
            local hil = "" .. tostring(self.state.s.highlights:len()) .. "❇"
            
            for i in iter(1, hil:len()) do 
                local color = (function () 
    if (i < hil:len()) then 
    return 'hil' else 
    return color.darken(theme.status.hil)
                        end
end)()
                add(hil[i], color, 'dark')
            end
        end
        
        -- fill to the right
        
        for ci = x, cols-1 do 
            add(' ', nil, 'dark')
        end
        
        add('', 'dark', self.color.gutter)
        
        return self:render()
    end

return status